### ISSUE TITLE
Widespread timeouts in Edge 17

### ISSUE THREAD

**Issue by foolip** [web-platform-tests/results-collection#563]
Edge results have always seemed to have a lot of TIMEOUT in the "Test file" row. Browsing around at random it's easy find cases for Edge 17 as well:
https://master-dot-wptdashboard.appspot.com/results/accelerometer/Accelerometer-enabled-by-feature-policy-attribute-redirect-on-load.https.html?sha=ad84337c03
https://master-dot-wptdashboard.appspot.com/results/compat/interfaces.any.html?sha=ad84337c03
https://master-dot-wptdashboard.appspot.com/results/wake-lock/wakelock-cancel-twice.https.html?sha=ad84337c03

@jugglinmike, as a start, could you count how many tests are like this, where only Edge has a harness-level timeout?

**Comment by gsnedders** [comment]
Do we know if this is Sauce specific or happens with native runs?

**Comment by foolip** [comment]
@thejohnjansen have you been able to compare the results from Sauce with your own runs yet?

**Comment by foolip** [comment]
I'm seeing this in https://wpt.fyi/results/old-tests/submission/Microsoft/selection?sha=eea0b54014&label=stable now.

**Comment by foolip** [comment]
Seeing this today also, in https://wpt.fyi/results/webmessaging/with-options?sha=c0a1bc68d7&label=stable

**Comment by foolip** [comment]
As a point of comparison, the same tests from the experimental Edge 18 run:
https://wpt.fyi/results/accelerometer/Accelerometer-enabled-by-feature-policy-attribute-redirect-on-load.https.html?sha=548797a187
https://wpt.fyi/results/compat/interfaces.any.html?sha=548797a187 (missing, not sure why)
https://wpt.fyi/results/wake-lock/wakelock-cancel-twice.https.html?sha=548797a187

The tests also run and have results when testing Edge 17 manually, so I'm quite sure this isn't a problem with Edge, but with the running infrastructure.

**Comment by foolip** [comment]
This issue became apparent again today on a [web-platform-tests that fail only in Chrome (from wpt.fyi data)](https://groups.google.com/a/chromium.org/d/msg/blink-dev/Met12DiCxVE/sQfo5UHQBgAJ) thread on blink-dev.

**Comment by foolip** [comment]
This may be the most extreme bad run I've spotted so far: https://wpt.fyi/results/?sha=5acd3bcf66

0 / N almost down the line. @jugglinmike, perhaps the setup has become totally broken somehow?

@lukebjerring, do you have an issue for "validate results"? Seems like with that in place we would reject a run like this, or label it bad at least.

**Comment by jugglinmike** [comment]
Well, the builds certainly aren't completing any faster. They're just slow *and* wrong now.

We're seeing three different kinds of errors. The first two:

    ▶ ERROR [expected OK] /css/css-scoping/slot-non-html-display-value.html
    └   → Message: Unable to get property 'addEventListener' of undefined or null reference (WARNING: The server did not provide any stacktrace information)

    ▶ ERROR [expected OK] /websockets/interfaces/WebSocket/bufferedAmount/bufferedAmount-large.html
    └   → Message: Unable to get property 'addEventListener' of undefined or null reference (WARNING: The server did not provide any stacktrace information)

...are always followed by a timeout, e.g.

    Command duration or timeout: 390 milliseconds
    Build info: version: '2.52.0', revision: '4c2593c', time: '2016-02-11 19:06:42'
    System info: host: 'sauce-win10', ip: '10.100.19.211', os.name: 'Windows 10', os.arch: 'amd64', os.version: '10.0', java.version: '1.8.0_131'
    Driver info: org.openqa.selenium.edge.EdgeDriver
    Capabilities [{applicationCacheEnabled=true, InPrivate=false, pageLoadStrategy=normal, platform=ANY, acceptSslCerts=true, browserVersion=42.17134.1.0, platformVersion=10, locationContextEnabled=true, webStorageEnabled=true, browserName=MicrosoftEdge, takesScreenshot=true, takesElementScreenshot=true, platformName=windows}]
    Session ID: XXXXXXXX-XXXX-XXXX-XXXX-XXXXE993DD85
    Stacktrace:
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0 (None:-2)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance (None:-1)
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance (None:-1)
        at java.lang.reflect.Constructor.newInstance (None:-1)
        at org.openqa.selenium.remote.ErrorHandler.createThrowable (ErrorHandler.java:206)
        at org.openqa.selenium.remote.ErrorHandler.throwIfResponseFailed (ErrorHandler.java:158)
        at org.openqa.selenium.remote.RemoteWebDriver.execute (RemoteWebDriver.java:678)
        at org.openqa.selenium.remote.RemoteWebDriver.executeAsyncScript (RemoteWebDriver.java:598)
        at sun.reflect.NativeMethodAccessorImpl.invoke0 (None:-2)
        at sun.reflect.NativeMethodAccessorImpl.invoke (None:-1)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke (None:-1)
        at java.lang.reflect.Method.invoke (None:-1)
        at org.openqa.selenium.support.events.EventFiringWebDriver$2.invoke (EventFiringWebDriver.java:103)
        at com.sun.proxy.$Proxy1.executeAsyncScript (None:-1)
        at org.openqa.selenium.support.events.EventFiringWebDriver.executeAsyncScript (EventFiringWebDriver.java:229)
        at org.openqa.selenium.remote.server.handler.ExecuteAsyncScript.call (ExecuteAsyncScript.java:58)
        at java.util.concurrent.FutureTask.run (None:-1)
        at org.openqa.selenium.remote.server.DefaultSession$1.run (DefaultSession.java:176)
        at java.util.concurrent.ThreadPoolExecutor.runWorker (None:-1)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run (None:-1)
        at java.lang.Thread.run (None:-1)

The third kind of error is far more common but far less helpful:

    ▶ ERROR [expected OK] /websockets/interfaces/WebSocket/bufferedAmount/bufferedAmount-readonly.html
    └   → 1

It is never followed by a timeout.

As you can see, the logs are now formatted with the "GroupingFormatter" because [that was recently made the default behavior of the WPT CLI](https://github.com/web-platform-tests/wpt/pull/13468). As mentioned there, this format is unideal for continuous integration because it limits the amount of information available in situations like these.

[I have reinstated the previous default logging format for this project.](https://github.com/web-platform-tests/results-collection/commit/13f72b446a8174d79c9f19804b0bd444e2f75c09) This won't fix anything, but it should help us in future debugging attempts.  [I've updated WPT so that I'm notified of changes to WPT's `tools` directory](https://github.com/web-platform-tests/wpt/pull/13629), so hopefully I can stay on top of things like this in the future.

In the mean time, [I have also updated to the latest release of the "Sauce Connect Binary"](https://github.com/web-platform-tests/results-collection/commit/5f43c41d4e096abfbdc0ab90baa1cb3883679db5) in the off chance that ["Fixed file descriptor leak."](https://wiki.saucelabs.com/display/DOCS/2018/08/27/Announcing+Sauce+Connect+Proxy+4.5.1) has some bearing here.

**Comment by foolip** [comment]
Thanks for investigating Mike! Do you think it'd be worth trying more aggressive restarting, and perhaps more chunks, to mitigate breakage like this?

**Comment by jugglinmike** [comment]
> Do you think it'd be worth trying more aggressive restarting

The fact that this immediately interferes with almost every "chunk" makes me think that restarting will not resolve it, but it couldn't hurt to try (especially now that we're disabling Safari on Sauce Labs). [We'll see what effect this has on time-to-results](https://github.com/web-platform-tests/results-collection/commit/27c22db74aefb3e1ae93d34a76c7c9502ce48b64) (it may get ugly).

(I was ready to point out that these are "ERROR"s and not "FAILURE"s, but then I remembered [that doesn't actually matter](https://github.com/web-platform-tests/wpt/pull/10186).)

> perhaps more chunks

[For Edge, we're currently running WPT in 100 "chunks"](https://github.com/web-platform-tests/results-collection/blob/27c22db74aefb3e1ae93d34a76c7c9502ce48b64/src/master/master.cfg#L27-L33) which already feels a bit excessive. I can give it a shot if we run out of ideas, though.

**Comment by foolip** [comment]
Oh, if it's already 100 chunks, then scratch that idea. Seems like something is broken with Sauce or the sauce support in `./wpt run` then?

**Comment by jugglinmike** [comment]
We've completed a collection from Edge that involved restarting after every test failure. This increased build time from 15 hours to 24 hours, but it had no appreciable affect on the veracity of the results.

Strangely, I haven't been able to connect to Sauce Labs locally. I'm not going to prioritize fixing that yet for a few reasons:

- This is distinct from the behavior we're seeing on the workers
- We've seen our share of roving DNS issues, so this may resolve itself in time

In the interim, I can do ad-hoc testing from any of the worker machines; that's probably more authentic, anyway. ...just not while they're actively testing. Although Edge is now running 24/7, 10 of our machines do not run Sauce Labs tests, so this is still a viable option.

**Comment by jugglinmike** [comment]
The problem comes from the way cross-realm objects are interpreted by Edge/Edgedriver. The easiest way to see this is by comparing the results of the WebDriver "execute script" command which collects test results from the browser.

<details>
  <summary>Excerpt from logs of a test conducted on 2018-10-14</summary>

```json
{
  "screenshot": null,
  "between_commands": 0.1099998950958252,
  "suggestion_values": [],
  "request": {
    "args": [],
    "script": "var callback = arguments[arguments.length - 1];\nwindow.testdriver_callback = callback;\nwindow.process_next_event();\n"
  },
  "HTTPStatus": 200,
  "result": [
    "/css/css-typed-om/the-stylepropertymap/properties/white-space.html",
    "complete",
    [
      0,
      null,
      null,
      [
        [
          "'white-space' property",
          1,
          "'CSSKeywordValue' is not defined",
          "ReferenceError: 'CSSKeywordValue' is not defined\n   at Global code (http://web-platform.test:8000/css/css-typed-om/the-stylepropertymap/properties/resources/testsuite.js:23:1)"
        ]
      ]
    ]
  ],
  "suggestion": null,
  "duration": 0.015000104904174805,
  "path": "execute_async",
  "method": "POST",
  "statusCode": 0,
  "timelineSeconds": 9.593000173568726
}
```

</details>

<details>
  <summary>Excerpt from logs of a test conducted on 2018-10-17</summary>

```json
{
  "screenshot": null,
  "between_commands": 0.2969999313354492,
  "suggestion_values": [],
  "request": {
    "args": [],
    "script": "var callback = arguments[arguments.length - 1];\nwindow.opener.testdriver_callback = callback;\nwindow.opener.process_next_event();\n"
  },
  "HTTPStatus": 200,
  "result": {
    "0": "/speech-api/historical.html",
    "1": "complete",
    "2": {
      "0": 0,
      "1": null,
      "2": null,
      "3": {
        "0": {
          "0": "SpeechRecognitionError interface should not exist",
          "1": 0,
          "2": null,
          "3": null
        },
        "1": {
          "0": "webkitSpeechGrammar interface should not exist",
          "1": 0,
          "2": null,
          "3": null
        },
        "2": {
          "0": "webkitSpeechGrammarList interface should not exist",
          "1": 0,
          "2": null,
          "3": null
        },
        "3": {
          "0": "webkitSpeechRecognition interface should not exist",
          "1": 0,
          "2": null,
          "3": null
        },
        "4": {
          "0": "webkitSpeechRecognitionError interface should not exist",
          "1": 0,
          "2": null,
          "3": null
        },
        "5": {
          "0": "webkitSpeechRecognitionEvent interface should not exist",
          "1": 0,
          "2": null,
          "3": null
        }
      }
    }
  },
  "suggestion": null,
  "duration": 0.25,
  "path": "execute_async",
  "method": "POST",
  "statusCode": 0,
  "timelineSeconds": 53.45300006866455
}
```
</details>

This has interfered with results collection because [the wptrunner expects to receive a Python list](https://github.com/web-platform-tests/wpt/blob/b704e37ec97fe90b3a3d59c10f78c21907b5b576/tools/wptrunner/wptrunner/executors/base.py#L517), but the structure that comes from Edge is deserialized into a dict. (We were seeing an error message of simply, "1" because of [an oversight in wptrunner's error handling code](https://github.com/web-platform-tests/wpt/pull/13806).)

All of the logging data provided by Sauce Labs indicates that the same build of Edge and Edgedriver were used in both cases:

- [data for the "good" trial](https://app.saucelabs.com/tests/26b3f133ad694c8dba211ee434b87682/logs#30)
- [data for the "bad" trial](https://app.saucelabs.com/tests/f1b336ce41214607ac1568c026293495/logs#12)

And I couldn't find any relevant notes in [the Sauce Labs product announcement blog](https://wiki.saucelabs.com/display/DOCS/Product+Announcements+and+Release+Notes+Blog) or on [Edgedriver's website](https://docs.microsoft.com/en-us/microsoft-edge/webdriver).

It turns out that this was actually caused by a change to WPT: https://github.com/web-platform-tests/wpt/pull/13419 . I've submitted [a fix that will accomodate Edge's non-standard behavior](https://github.com/web-platform-tests/wpt/pull/13881) and [a separate patch with WebDriver specification tests](https://github.com/web-platform-tests/wpt/pull/13880).

**Comment by jugglinmike** [comment]
We landed a short-term fix for this in WPT:

https://github.com/web-platform-tests/wpt/pull/13938

Since that was merged today, the change will not influence collection until tomorrow. I'll leave this open until that can be verified (likely not until Thursday).

**Comment by foolip** [comment]
Still no results on wpt.fyi, what is the status of the Edge run that started after the fix?

**Comment by jugglinmike** [comment]
Good news first: [we successfully collected from Edge, the data appears to be valid, and the restart-enabled attempt took less than 24 hours](https://wpt.fyi/results/?sha=71fe0a6c3d).

The reason you don't see this on the frontpage of https://wpt.fyi is that the corresponding collection for Chrome failed. If it's not one browser...

I've manually triggered a retry, and filed gh-625 to track this latest instability.
